source/atomic_ops.cpp

```diff
@@ -63,7 +62,56 @@ bool atomic_cas<uint32_t>(uint32_t *ptr, uint32_t *expectedCurrentValue, uint32_
 
     return !__STREXW(desiredValue, ptr);
 }
-#endif /* #if (__CORTEX_M >= 0x03) */
+
+uint32_t atomic_incr(uint32_t *valuePtr, uint32_t delta)
+{
+    uint32_t newValue;
+    do {
+        newValue = __LDREXW(valuePtr) + delta;
```

Comment:
it would be better to have platform specific atomic-access operations hidden behind a single API. atomic_cas serves that purpose. Algorithms requiring atomicity should be built upon this primitive. There is no need to provide other parallel ways of achieving atomicity. 

This whole discussion began when we unearthed a flaw in our current implementation of template based atomics. There is a simple fix for it--add missing declarations in the header. This problem can also be solved using the proposed rework which translates template specializations to overloads, and that may not be such a bad thing. I certainly object to the use of platform-specific primitives like LDREX in atomic_incr. But I'm not going to fight this any further. If others are aligned with this proposal, then go ahead.